sdks/java/extensions/smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -0,0 +1,304 @@
+
+            valueMap.put(entry.getKey(), values);
+            nextKeyGroupsIt.remove();
+          }
+        }
+
+        c.output(KV.of(keyForGroup, toResult.apply(new SMBCoGbkResult(valueMap))));
+
+        if (readers.isEmpty()) {
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Maintains type information about a possibly heterogeneous list of sources by wrapping each one
+   * in a KeyedBucketSource object with TupleTag. Heavily copied from CoGroupByKey implementation.
+   *
+   * @param <K>
+   */
+  public static class KeyedBucketSources<K>
+      implements Serializable, org.apache.beam.sdk.values.PInput {
+    private transient Pipeline pipeline;
+    private List<KeyedBucketSource<K, ?>> sources;
+    private Integer numBuckets;
+
+    KeyedBucketSources(Pipeline pipeline, Integer numBuckets) {
+      this(pipeline, numBuckets, new ArrayList<>());
+    }
+
+    KeyedBucketSources(
+        Pipeline pipeline, Integer numBuckets, List<KeyedBucketSource<K, ?>> sources) {
+      this.pipeline = pipeline;
+      this.numBuckets = numBuckets;
+      this.sources = sources;
+    }
+
+    static <K, InputT> KeyedBucketSources<K> of(
+        Pipeline pipeline, int numBuckets, KeyedBucketSource<K, InputT> source) {
+      return new KeyedBucketSources<K>(pipeline, numBuckets).and(source);
+    }
+
+    <V> KeyedBucketSources<K> and(KeyedBucketSource<K, V> source) {
+      List<KeyedBucketSource<K, ?>> newKeyedCollections = copyAddLast(sources, source);
+
+      return new KeyedBucketSources<>(pipeline, numBuckets, newKeyedCollections);
+    }
+
+    @Override
+    public Pipeline getPipeline() {
+      return pipeline;
+    }
+
+    @Override
+    public Map<TupleTag<?>, PValue> expand() {
+      final Map<TupleTag<?>, PValue> map = new HashMap<>();
+
+      final List<KV<Integer, List<BucketSourceIterator<K>>>> readers = new ArrayList<>();
+      for (int i = 0; i < numBuckets; i++) {
+        try {
+          readers.add(KV.of(i, createReaders(i)));
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+      final Coder<KV<Integer, List<BucketSourceIterator<K>>>> coder =
+          KvCoder.of(VarIntCoder.of(), ListCoder.of(new BucketSourceIteratorCoder<K>()));
+
+      map.put(
+          new TupleTag<K>("readers"),
+          pipeline.apply("Create bucket readers", Create.of(readers).withCoder(coder)));
+
+      return map;
+    }
+
+    private static <K> List<KeyedBucketSource<K, ?>> copyAddLast(
+        List<KeyedBucketSource<K, ?>> keyedCollections, KeyedBucketSource<K, ?> taggedCollection) {
+      final List<KeyedBucketSource<K, ?>> copy = new ArrayList<>(keyedCollections);
+      copy.add(taggedCollection);
+      return copy;
+    }
+
+    /**
+     * Represents a single source with values V.
+     *
+     * @param <K>
+     * @param <V>
+     */
+    public static class KeyedBucketSource<K, V> implements Serializable {
```

Comment:
Can get rid of type param `K` here and `createReaders` below if it returns byte array as key.